### PR TITLE
Sections view

### PR DESCRIPTION
**Pull request by hn275**
# Section View

Changes to selected sections are reflected in the set cookie for the term.

## UI

screenshot

### CODE REVIEW

File: src/components/container.rs [hunk 1]
```diff
@@ -38,11 +44,45 @@ pub fn calendar_view_container(oob: bool) -> Markup {
...
+                                div class="flex gap-2 justify-between items-center" {
+                                    h3 {
+                                        (&course.subject_code) " " (&course.course_code) " - " (&course.title)
+                                    }
+                                    form {
+                                        input type="hidden" name="subject_code" value=(course.subject_code){}
+                                        input type="hidden" name="course_code" value=(course.course_code){}
+                                        button name="course" value={(course.subject_code) " " (course.course_code)}
+                                        class="bg-red-500 dark:bg-red-600 hover:bg-red-700 hover:dark:bg-red-800 text-black dark:text-white rounded-md p-1 shadow-lg"
+                                        hx-delete={"/term/" (term) "/calendar"} hx-swap="none" {
```
Comment (brennanmcmicking) [inline src/components/container.rs:74]: no swap? this should swap the entire calendar view as far as i'm aware

File: src/components/container.rs [hunk 2]
```diff
@@ -38,11 +44,45 @@ pub fn calendar_view_container(oob: bool) -> Markup {
...
+                        details class="py-3" {
+                            summary class="flex cursor-pointer" {
+                                div class="flex gap-2 justify-between items-center" {
+                                    h3 {
+                                        (&course.subject_code) " " (&course.course_code) " - " (&course.title)
+                                    }
+                                    form {
+                                        input type="hidden" name="subject_code" value=(course.subject_code){}
+                                        input type="hidden" name="course_code" value=(course.course_code){}
+                                        button name="course" value={(course.subject_code) " " (course.course_code)}
```
Comment (brennanmcmicking) [inline src/components/container.rs:72]: what does this make the request body look like when you click the button?

File: src/components/section_view.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        }
+    }
+}
+
+fn checkbox(url: &str, crn: u64, checked: bool) -> Markup {
+    // bc I can't figure out how to inline the `checked` html
+    // attribute properly :(
+    match checked {
+        true => html! {
+            input type="checkbox"
```
Comment (brennanmcmicking) [inline src/components/section_view.rs]: this should be type radio, not checkbox. you can't select multiple lab sections at the same time

File: src/components/section_view.rs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                hx-put=(url)
+                name="crns"
+                value=(crn);
+        },
+    }
+}
+
+fn meeting_time(meeting_time: &[MeetingTime]) -> Markup {
+    // I'm not sure why this meeting time has to be a vec
+    let meeting_time = &meeting_time[0];
```
Comment (brennanmcmicking) [inline src/components/section_view.rs]: there seems to be a newline between each statement, can you please clean up the code as much as possible?

File: src/components/section_view.rs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        let pattern = "%I:%M %p"; // formats to 03:00 PM
+        let start = &meeting_time.start_time.unwrap().strftime(pattern);
+        let end = &meeting_time.end_time.unwrap().strftime(pattern);
+        format!("{} - {}", start, end)
+    } else {
+        String::from("N/A")
+    };
+
+    html! {
+        div class="flex justify-between items-center text-sm gap-2 w-full"{
```
Comment (brennanmcmicking) [inline src/components/section_view.rs]: missing space before the `{` here (and other places)

File: src/components/section_view.rs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                            (meeting_time(&section.meeting_times))
+
+                            div class="flex justify-between items-center w-full" {
+
+                                span class="text-sm" {
+                                    "Seats: " (section.enrollment)"/"(section.enrollment_capacity)
+                                }
+
+                                span class="text-sm" {
+                                    "Waitlist: " (section.waitlist)"/"(section.waitlist_capacity)
```
Comment (brennanmcmicking) [inline src/components/section_view.rs]: maybe only show waitlist count if the enrolment is full?

File: src/components/section_view.rs (hunk 4 above)
Comment (brennanmcmicking) [inline src/components/section_view.rs]: you can't just use the first meeting time. some sections have multiple

File: src/routes.rs [hunk 7]
```diff
@@ -69,7 +69,90 @@ impl DatabaseAppState {
         terms
     }
 
-    fn courses(&self, term: Term) -> Result<Vec<ThinCourse>> {
+    fn courses(&self, term: Term, keys: &[&ThinCourse]) -> Result<Vec<Course>> {
```
Comment (brennanmcmicking) [inline src/routes.rs:72]: i'll leave @malcolmseyd to comment on the business logic here, i'm not really familiar with what this ideally would look like

File: src/routes.rs [hunk 8]
```diff
@@ -207,6 +290,10 @@ pub async fn make_app() -> Router {
                 .route(
                     "/calendar",
                     put(calendar::add_to_calendar).delete(calendar::rm_from_calendar),
+                )
+                .route(
+                    "/section",
+                    put(calendar::course_section),
```
Comment (brennanmcmicking) [inline src/routes.rs]: this function name isn't very descriptive on what it does

**Review by brennanmcmicking — CHANGES_REQUESTED** [review]

File: src/components/container.rs (hunk 2 above)
Comment (malcolmseyd) [reply src/components/container.rs:72]: There's actually no request body. HTMX encodes form values as URL parameters on DELETE requests to comply with the HTML spec.
https://github.com/bigskysoftware/htmx/blob/d152a3c75a8a2bd905379de068e6a79a7dddcca0/src/htmx.js#L230-L235

But I think I see what you're getting at: PUT and DELETE to `/term/:term/calendar` both only take a single parameter, `course`, which is a space-delimited string that parses to a `ThinCourse`.

https://github.com/brennanmcmicking/scheduler/blob/7bd800cea01173e3f1eaf9311800cfd96dadfdcf/src/routes/calendar.rs#L18-L29

We can remove the hidden form elements in both `search_result::render` and `container::courses_container` and it'll behave the exact same.

File: src/routes/calendar.rs [hunk 9]
```diff
@@ -25,61 +24,185 @@ pub struct Search {
...
         },
     ))
 }
+
+// HANDLER FOR COURSE SECTIONS
+// because Axum doesn't support duplicated fields in form data...
+#[derive(Debug, PartialEq)]
+struct SectionQuery(ThinCourse, Vec<ThinSection>);
+
+impl TryFrom<&RawForm> for SectionQuery {
```
Comment (malcolmseyd) [inline src/routes/calendar.rs]: Please use [this `Form` extractor instead](https://docs.rs/axum-extra/latest/axum_extra/extract/struct.Form.html#differences-from-axumextractform), it's supported alongside the official axum project.

File: src/components/container.rs (hunk 2 above)
Comment (hn275) [reply src/components/container.rs:72]: :+1: for url param, but I'm not sure we can remove these hidden fields for this, or at least I couldn't come up with a solution that would allow so. 

The raw formdata string is `course_code=xxx&section_code=xxx&crn=1&crn=2...`. All of the checkbox values in the form will be sent, (hence why the html it generates has a biiigggg form), the server just replaces the internal crn seletions with whatever that is sent to the server

I need the course code and section code to search for selected crns in user state cookie.

I'd be happy to re-implement this if you guys can think of a sane solution

File: src/components/section_view.rs (hunk 4 above)
Comment (hn275) [reply src/components/section_view.rs]: ~~I've actually never seen a section with multiple meeting time in the json we get back from banner, I think they just have it in an array by default? I should check for empty array though. maybe @malcolmseyd can confirm~~

maybe this is more common in non-stem majors, I'm just going to make it render all meeting times

File: src/components/section_view.rs (hunk 6 above)
Comment (hn275) [reply src/components/section_view.rs]: but then you'd have blank white space in the section card for those which aren't full, it looks kinda imbalance imo

File: src/components/section_view.rs (hunk 4 above)
Comment (hn275) [reply src/components/section_view.rs]: I don't know if this is better

```rs
    let meeting_time = &meeting_time[0];
    let days = &meeting_time.days;
    let mut day_str = String::new();
    if days.monday {
        day_str.push('M');
    }
    if days.tuesday {
        day_str.push('T');
    }
    if days.wednesday {
        day_str.push('W');
    }
    if days.thursday {
        day_str.push_str("Th");
    }
    if days.friday {
        day_str.push('F');
    }
```
the homies with modal text editor may appreciate the newlines.

File: src/components/container.rs (hunk 1 above)
Comment (hn275) [reply src/components/container.rs:74]: you're right, this would be related to what you're working on in #27 i think? Do I swap with a `<div id="calendar">`? Mind you right now there's no data being sent back, (other than updating the cookie). would it be easier to come back for this once #27 is pulled?

File: src/components/section_view.rs (hunk 4 above)
Comment (brennanmcmicking) [reply src/components/section_view.rs]: engr 110 lectures at least used to have multiple meetings times

**Comment by hn275** [comment]
this will also close #24

File: src/components/section_view.rs (hunk 6 above)
Comment (brennanmcmicking) [reply src/components/section_view.rs]: why would you have blank whitespace for sections which aren't full?

File: src/components/section_view.rs (hunk 4 above)
Comment (brennanmcmicking) [reply src/components/section_view.rs]: that's not really a sufficient justification in my mind for random whitespace. please elaborate.

also please don't mark conversations as resolved until all people involved have agreed

File: src/components/container.rs (hunk 1 above)
Comment (brennanmcmicking) [reply src/components/container.rs:74]: i'm not sure cause i don't have my laptop on me. maybe ask malcolm

File: src/components/container.rs [hunk 10]
```diff
@@ -38,11 +44,44 @@ pub fn calendar_view_container(oob: bool) -> Markup {
...
+                @for course in courses {
+                    div id="courses-card" class="rounded-lg flex-col justify-center items-center p-1 dark:text-white border border-black dark:border-white" {
+                        details class="py-3" {
+                            summary class="flex cursor-pointer" {
+                                div class="flex gap-2 justify-between items-center" {
+                                    h3 {
+                                        (&course.subject_code) " " (&course.course_code) " - " (&course.title)
+                                    }
+                                    form {
+                                        input type="hidden" name="subject_code" value=(course.subject_code){}
```
Comment (brennanmcmicking) [inline src/components/container.rs:70]: missing space here before {}

File: src/components/container.rs (hunk 2 above)
Comment (brennanmcmicking) [reply src/components/container.rs:72]: that doesn't make any sense, why do you need you search for crns to delete a course

File: src/components/section_view.rs [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+        .map(|s| s.crn)
+        .collect::<Vec<_>>();
+
+    let url = format!("/term/{}/section", term);
+
+    html! {
+        form class="px-3" {
+            input type="hidden" name="course" value=(format!("{} {}",&course.subject_code, &course.course_code));
+
+            (section( "lectures",&url, &lectures, &selected_crns))
```
Comment (brennanmcmicking) [inline src/components/section_view.rs:51]: unnecessary whitespace after `section(`

File: src/components/section_view.rs [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+        }
+    }
+}
+
+fn checkbox(url: &String, name: &str, crn: u64, checked: bool) -> Markup {
+    // bc I can't figure out how to inline the `checked` html
+    // attribute properly :(
+    match checked {
+        true => html! {
+            input type="radio"
```
Comment (brennanmcmicking) [inline src/components/section_view.rs:117]: did you test this out locally? it seems like you'd still be able to check multiple at the same time with this code

File: src/components/section_view.rs [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+            (section( "lectures",&url, &lectures, &selected_crns))
+            (section( "labs",&url, &labs, &selected_crns))
+            (section( "tutorials",&url, &tutorials, &selected_crns))
+
+        }
+    }
+}
+
+fn section(
+    section_type: &str,
```
Comment (brennanmcmicking) [inline src/components/section_view.rs:60]: create an enum for this please, we're living in OOP land

File: src/components/section_view.rs [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn checkbox(url: &String, name: &str, crn: u64, checked: bool) -> Markup {
+    // bc I can't figure out how to inline the `checked` html
+    // attribute properly :(
```
Comment (brennanmcmicking) [inline src/components/section_view.rs:114]: src/components/container.rs line 59 in this pr may be what you're looking for

File: src/middlewares.rs [hunk 15]
```diff
@@ -17,6 +17,10 @@ pub struct SelectedCourses {
 }
 
 impl SelectedCourses {
+    pub fn thin_courses(&self) -> Vec<&ThinCourse> {
```
Comment (brennanmcmicking) [inline src/middlewares.rs:20]: please put non-static functions below static functions. a little organization goes a long way for navigability

File: src/routes/calendar.rs [hunk 16]
```diff
@@ -25,61 +24,96 @@ pub struct Search {
...
         jar,
         html! {
             (calendar_view_container(true))
-            (courses_container(true))
+            (courses_container(true, term, &courses, &new_selected))
         },
     ))
 }
+
+// HANDLER FOR COURSE SECTIONS
```
Comment (brennanmcmicking) [inline src/routes/calendar.rs:96]: WHY ARE WE YELLING

File: src/routes/calendar.rs [hunk 17]
```diff
@@ -25,61 +24,96 @@ pub struct Search {
...
+#[derive(Debug, PartialEq, Deserialize)]
+pub struct SectionQuery {
+    course: ThinCourse,
+
+    #[serde(default)] // if no box checked, field does not exist
```
Comment (brennanmcmicking) [inline src/routes/calendar.rs:101]: there should never be no box checked. when you add a course the first section of each type (lecture, lab, tutorial) should be selected

Comment (brennanmcmicking) [reply src/routes/calendar.rs:101]: also it's not a box it's a radio button now

File: src/routes/calendar.rs [hunk 18]
```diff
@@ -25,61 +24,96 @@ pub struct Search {
...
+    crns: Vec<u64>,
+}
+
+#[instrument(level = "debug")]
+pub async fn course_section_handler(
+    Path(term): Path<Term>,
+    mut selected: SelectedCourses,
```
Comment (brennanmcmicking) [inline src/routes/calendar.rs:108]: ~~i'm not liking the look of this being mut, i suspect there's a better way to do things~~

meh nvm it's fine

File: src/routes/calendar.rs [hunk 19]
```diff
@@ -25,61 +24,96 @@ pub struct Search {
...
+    ExtractedForm(form): ExtractedForm<SectionQuery>,
+) -> Result<impl IntoResponse, AppError> {
+    let sections = form
+        .crns
+        .iter()
+        .map(|&crn| ThinSection { crn })
+        .collect::<Vec<_>>();
+
+    selected.courses.insert(form.course, sections);
+    Ok(CookieJar::new().add(selected.make_cookie(term)))
```
Comment (brennanmcmicking) [inline src/routes/calendar.rs:118]: this will also trigger a render, shouldn't it return html?

**Review by brennanmcmicking — CHANGES_REQUESTED** [review]

File: src/components/container.rs (hunk 1 above)
Comment (malcolmseyd) [reply src/components/container.rs:74]: Use oob swap, that's what i did with the other components

File: src/components/section_view.rs (hunk 4 above)
Comment (hn275) [reply src/components/section_view.rs]: I don't think it's random, in my mind each of the whitespace is doing a thing separately, but it's your code base so I'm not going to argue.

I marked them resolved since newer commits made the requested change and the conversation became outdated

File: src/components/container.rs (hunk 2 above)
Comment (hn275) [reply src/components/container.rs:72]: apparently I was thinking of something else entirely, ignore me

File: src/routes.rs (hunk 7 above)
Comment (malcolmseyd) [reply src/routes.rs:72]: I wrote this so LGTM :laughing: :+1:

File: src/components/section_view.rs (hunk 4 above)
Comment (malcolmseyd) [reply src/components/section_view.rs]: Yeah if a commit marks it as outdated, and you think the resolution is pretty clear, feel free to resolve without comment :+1:.

If you're not sure if the commit fully addressed the feedback, or you think there's still discussin to be had, feel free to link the commit and explain your change, but I don't think this is always necessary.

File: src/components/container.rs (hunk 2 above)
Comment (malcolmseyd) [reply src/components/container.rs:72]: I think I see what you might have been going for. In my mind there are two options:

- The user clicks a radio button and triggers form resubmission for just that button group. This acts as a diff, where for example, the lecture section's CRN is replaced with the one submitted. This puts the idea of "section types" into the handler code, as logic that requires database lookups (we need crn -> sequence_code to do this).
- The user clicks a radio button and triggers form resubmission for all of that course's section types. This acts as a full replacement, and all selected radio buttons are submitted to replace all previous state for that course with, simply, all currently selected radio buttons. This makes the submission handler simpler, requiring no additional database lookups (just blindly replace the state for that course with listed CRNs), and moves the grouping logic to the view. In the view we already have seqence_codes (required for grouping radio buttons in UI by i.e. A01, A02, B01, B01). The downside is it provides less validation in the backend that submitted data is "correct" (i.e. someone selects two lecture sections in devtools somehow).

Both of these options are totally fine imo, and I'm curious to hear both of your opinions on this @brennanmcmicking @hn275. Feel free to implement either approach, but I think it's important to document why we might choose either approach, and what goes into that decision, at least in this PR thread to refer back to later.

**Comment by brennanmcmicking** [comment]
Some general feedback after testing this out locally on my laptop:
* The "remove" button isn't centered properly
* The area to select sections for a course are by-default collapsed and it's not obvious to a new user how to find them (it took me a non-zero amount of time to figure it out)
* Picking a new selection makes a PUT request, not a PATCH request. Adding a course should PUT the course into the `/.../calendar` route, picking a new section should PATCH the new section ID (either CRN or course name + sequence number i.e. "MATH 100 A01") to `/.../calendar`, and removing a course should DELETE it from `/.../calendar`
* IMO too much liberty was taken from the [existing UI mockup](https://excalidraw.com/#json=Gy2QfYj48tbT_JeTS-Lju,s66OheueOnmxZdSqZ1CQsw)
* The section view looks crowded with a lot of information that is poorly-organized and superficial (why show the CRN?)

This is not meant to be a criticism, but simply my perception of a "lack of polish" that, if addressed, will take this PR from "okay" to "amazing"
